components/psst/browser/content/psst_tab_helper.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/psst/browser/content/psst_tab_helper.h"
+
+#include <string>
+#include <utility>
+
+#include "base/functional/bind.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/thread_pool.h"
+#include "base/values.h"
+#include "brave/components/psst/browser/core/psst_rule.h"
+#include "brave/components/psst/browser/core/psst_rule_service.h"
+#include "brave/components/psst/common/features.h"
+#include "components/sessions/content/session_tab_helper.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "mojo/public/cpp/bindings/associated_remote.h"
+#include "third_party/blink/public/common/associated_interfaces/associated_interface_provider.h"
+
+namespace psst {
+
+// static
+void PsstTabHelper::MaybeCreateForWebContents(content::WebContents* contents,
+                                              const int32_t world_id) {
+  if (contents->GetBrowserContext()->IsOffTheRecord() ||
+      !base::FeatureList::IsEnabled(psst::features::kBravePsst)) {
+    return;
+  }
+
+  psst::PsstTabHelper::CreateForWebContents(contents, world_id);
+}
+
+PsstTabHelper::PsstTabHelper(content::WebContents* web_contents,
+                             const int32_t world_id)
+    : WebContentsObserver(web_contents),
+      content::WebContentsUserData<PsstTabHelper>(*web_contents),
+      world_id_(world_id),
+      psst_rule_service_(PsstRuleService::GetInstance()) {
+  DCHECK(psst_rule_service_);
+}
+
+PsstTabHelper::~PsstTabHelper() = default;
+
+void PsstTabHelper::OnTestScriptResult(
+    const std::string& policy_script,
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    base::Value value) {
+  if (value.GetIfBool().value_or(false)) {
+    InsertScriptInPage(render_frame_host_id, policy_script, base::DoNothing());
+  }
+}
+
+void PsstTabHelper::InsertTestScript(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    MatchedRule rule) {
+  CHECK(render_frame_host_id);
```

Comment:
I didn't notice this before, but this check doesn't make sense to me. You're asserting true/false on something that is not a true/false or zero/nonzero value because it's a struct. This couldn't fail in practice.

Reply:
oh wait, I see now that `GlobalRenderFrameHostId` implements `operator bool` which is why this compiles, but I still don't think it makes sense here as a check because the frame routing id won't change from the time you get the `GlobalRenderFrameHostId`. If anything this should just be an early return if you've somehow passed an invalid `GlobalRenderFrameHostId`